firebase_test.go

```diff
@@ -338,6 +338,224 @@ func TestVersion(t *testing.T) {
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	app, err := NewApp(context.Background(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := &Config{
+		DatabaseURL:   "",
+		ProjectID:     "mock-project-id", // from default credentials
+		StorageBucket: "",
+	}
+	compareConfig(app, want, t)
+}
+
+func TestAutoInitPartialOverride(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/firebase_config_partial.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	app, err := NewApp(context.Background(),
+		&Config{
+			DatabaseURL:   "db1-mock",
+			StorageBucket: "sb1-mock",
+		})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	want := &Config{
+		DatabaseURL:   "db1-mock",
+		ProjectID:     "hipster-chat-mock",
+		StorageBucket: "sb1-mock",
+	}
+	compareConfig(app, want, t)
+}
+
+func TestAutoInitPartialOverrideWithoutEnv(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/firebase_config_partial.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+	os.Unsetenv(FirebaseEnvName)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	app, err := NewApp(context.Background(),
+		&Config{
+			DatabaseURL: "db1-mock",
+			ProjectID:   "pid1-mock",
+		})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	want := &Config{
+		DatabaseURL: "db1-mock",
+		ProjectID:   "pid1-mock",
+	}
+	compareConfig(app, want, t)
+}
+
+func TestAutoInitNoFile(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/no_such_file.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	_, err := NewApp(context.Background(), &Config{})
+
+	we := "open testdata/no_such_file.json: no such file or directory"
+	if err == nil || err.Error() != we {
+		t.Errorf("got error = %s; wanted %s", err, we)
+	}
+}
+
+func TestAutoInitNoFileButNotNeeded(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/no_such_file.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	we := "open testdata/no_such_file.json: no such file or directory"
+
+	_, err := NewApp(context.Background(), &Config{DatabaseURL: "d", ProjectID: "p", StorageBucket: "s"})
```

Comment:
Is this the behavior we want? I thought we implemented different semantics in Node?

Reply:
deleted in a previous commit.